### PR TITLE
[Form] [Component] Add support for an initial form value

### PR DESCRIPTION
**Pull request by Bossman1337**
| Q             | A
| ------------- | ---
| Branch?       | 6.1 
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| License       | MIT

Add support for an **_initial form value_** to be set when the form does not have an underlying object (yet) or not bound.

There are many times i see confusion over the `data` and `empty_data` properties in form types. These do not provide the expected functionality that many people think, especially new comers. It is more intuitive to have an `init_value` (initial value).

Even though this functionality can be achieved via a form listener, it should be a default feature. It can also be achieved via a placeholder, but this can differ depending on front-end CSS frameworks used that may need modifying. Again, it is more intuitive with `init_value`.

 **Example usuage:**

```
$builder
    ->add('field_name', TextType::class, array(
        'init_value' => 'Hello World!'
));
```
**TODO:**

- [ ] Modify/Update documentation if this PR gets accepted.

### CODE REVIEW

**Review by yceruto — COMMENTED** [review]
There is already a way to set the default value to any form field and it's precisely through the underlying data (object or array). And for unmapped form fields, you can safely use `data` option directly.

Can you elaborate a bit more on why you cannot bind it to the form?

**Comment by Bossman1337** [comment]
@yceruto Hi, 

If you have an entity and would like to _insert/edit_ some data via a form, you would have the form bound to that entity.
Now, if you would like a field to have a "default value" populated in that field when the form is loaded, this is what `init_value` does. When you come to edit the entity via the form, it will pull the value from the database as it exists.

`data` does not work for this scenario and is explained in the documentation [here](https://symfony.com/doc/current/reference/forms/types/text.html#data).

In effect it is like doing `attr => ['value' = 'my value']` this won't work either because if there is a value for this field in the database, this will overwrite that.

Basically, if there is a value set in the database it will use that, if not it will use `init_value`.

This is my point, everyone always says use `data` or `empty_data`.

**Comment by yceruto** [comment]
@Bossman1337 Maybe the documentation is not completely clear about this subject, but the default values should be set directly in the underlying data before binding it to the form. The `data` and `empty_data` options have different meanings and they are well documented.

Some examples in the Form documentation show how to do it but I guess it's not explicitly written. When you're creating a new object/document (DTO or Entity) you should set the default values during its initialization process:
```php
$task = new Task();
$task->setReporter($user); // default value
$task->setPriority(0); // default value
// more default values...

$form = $formFactory->create(TaskType::class, $task);
```
And for those underlying data that are intended to be edited, you probably don't want to set a default value, but in case you have to do it, you still can before binding the Form.

> Basically, if there is a value set in the database it will use that, if not it will use init_value.

I can't say for sure that because a field data is `null` I want to set always a default value, sometimes `null` is the value that I want when I'm editing an object/document with optional properties; if I want to set a default value only when the object is being created for the first time then I can't reuse the same Form Type for editing, or maybe I want to set a default value if the field data is `false` or empty string `''`. 

It's not only one default mechanism, but it also depends on your use case, and that's why I believe this logic shouldn't be included in the Form context.

**Comment by Bossman1337** [comment]
@yceruto, Hi,

I understand and know where you're coming from, there are cases where it is useful for this to live in the form and not set in the controller (eg. dynamic form collections).

> I can't say for sure that because a field data is null I want to set always a default value, sometimes null is the value that I want when I'm editing an object/document with optional properties; if I want to set a default value only when the object is being created for the first time then I can't reuse the same Form Type for editing, or maybe I want to set a default value if the field data is false or empty string ''.

You can omit `init_value`, its optional, the forms will work as normal. You can reuse the form editing, as the data will be set and it will use that from now on, it won't use the `init_value` if there is data.

Like you say maybe the docs just need improving.

IMO i think it is a useful feature. If it doesn't get added then it doesn't get added, i'll continue to use it in projects.

**Comment by stof** [comment]
@Bossman1337 for collections, there is already a `prototype_data` option allowing to control the data set in the prototype (which would then prefill fields added based on the prototype)

**Comment by Bossman1337** [comment]
@stof, yes that is correct. However without going into detail, we use our own custom generated prototypes, this was an easy to understand solution in our context.

**Comment by stof** [comment]
@Bossman1337 well, then use a custom solution as well (or make your custom prototypes respect the `prototype_data` option) instead of adding a new feature in Symfony itself.

**Comment by Bossman1337** [comment]
@stof thats fair, and we currently are doing so, just thought it would be a good core feature that would benefit others and make it more intuitive. Like i say, if its not accepted then so be it..

**Comment by stof** [comment]
Well, as explained by @yceruto, this feature is actually worse than the current way of initializing the data (as it assumes `null` means uninitialized, which is wrong for nullable properties that you might edit) and we already have a feature in core for the case of collections. Adding a new way that is inferior to the existing one (and then having to maintain both and requesting users to choose between both) is a bad idea to me.

this gets a -1 vote from me.

**Comment by Bossman1337** [comment]
@stof If a property is null it will return null anyways as `init_value` default is null.

> as it assumes null means uninitialized, which is wrong for nullable properties that you might edit.

Its a chicken and egg situation. If you used `init_value` to begin with then the edit will show this. Like i mentioned previously, it is just setting the rendered value on the form.

**Comment by stof** [comment]
@Bossman1337 I'm talking about cases where your `init_value` option has something else than null (the case of not using `init_value` is not the one causing issue, but that case is also not the one justifying to add the feature)

**Comment by Bossman1337** [comment]
@stof 

> I'm talking about cases where your init_value option has something else than null

Yes, then in that case it wont be used as there is already data. It's prioritising the data already in the database.

**Comment by stof** [comment]
@Bossman1337 If you consider that a `null` value in the object is a case of "there is already data", then `init_value` is never used, making it even less useful. And if you consider that a `null` value in the object should use the `init_value`, then that's precisely where both @yceruto and me say that this is broken, as you might have `null` in the database.

**Comment by Bossman1337** [comment]
@stof Okay, obviously i am doubting myself now, maybe a different way of thinking..

Lets say you have an entity: `Customer`. It has 20 fields.

- The client has a form where they can enter values and insert a customer.
- The client does this daily adding 100 customers a day.
- The client asks you; "90% of the time when adding a new customer, 15 of these fields are the same value, can you make the form pre-populate these 15 fields with these values so it is quicker for us to add a customer?"

There are multiple ways to achieve this, you could do this through the controller passing the options or you could do this within the form builder via `init_value` (initial rendered value, brand new insert, not used when editing), where _IMO_, this should live.

Thanks for your time anyways, hope i'm not annoying you too much..

**Comment by stof** [comment]
@Bossman1337 doing that in the form builder as you say would require that your form type is only used for inserting, not for editing. That's the edge case to me (in most cases, your edition page needs to be consistent with your creation page).

**Comment by fabpot** [comment]
Closing for the reasons explained in the conversation.
